File: src/file_io.rs
use std::{
    error::Error,
    fs::remove_file,
    path::PathBuf,
};

use crate::save_model::{self, Character};

pub fn list_active_characters(data: &[u8]) -> Vec<Character> {
    (0..10)
        .filter_map(|index| Character::new_active(data, index))
        .collect()
}

pub fn list_characters(data: &[u8]) -> Vec<Option<Character>> {
    (0..10)
        .map(|index| Character::new_active(data, index))
        .collect()
}

/// Prases all chraracters, active or inactive
pub fn list_all_characters(data: &[u8]) -> Vec<Character> {
    (0..10).map(|i| Character::new(data, i)).collect()
}


pub fn write_file(data: &[u8], fully_qualified_file_name: &PathBuf) -> Result<(), Box<dyn Error>> {
    std::fs::write(fully_qualified_file_name, data)?;
    Ok(())
}

fn subslice_positions(needle: &[u8], haystack: &[u8]) -> Vec<usize> {
    haystack
        .windows(needle.len())
        .enumerate()
        .filter(|(_, slice)| slice[..] == needle[..])
        .map(|(i, _)| i)
        .collect()
}

// We are taking owenership because the upstream does it
// Might address this  later if there is a reason for it
pub fn generate_new_data(
    source_data: &[u8],
    source_character_slot: usize,
    target_data: &[u8],
    target_character_slot: usize,
) -> Result<Vec<u8>, Box<dyn Error>> {
    let mut new_save = target_data.to_owned();

    let source_id = save_model::parse_steam_id(source_data)?;
    let target_id = save_model::parse_steam_id(target_data)?;

    let mut source_character_save_data =
        save_model::parse_save_data(source_data, source_character_slot);

    let source_character_header_data =
        save_model::parse_header_data(source_data, source_character_slot);

    for id_location in subslice_positions(&source_id, &source_character_save_data) {
        source_character_save_data[id_location..id_location + save_model::STEAM_ID_LENGTH]
            .copy_from_slice(&target_id);
    }

    // Copy source save slot to target save slot in temp file
    new_save[save_model::get_slot_start_position(target_character_slot)..]
        [..save_model::SLOT_LENGTH]
        .copy_from_slice(&source_character_save_data);

    // Copy save header to temp file
    new_save[save_model::get_header_start_position(target_character_slot)..]
        [..save_model::SAVE_HEADER_LENGTH]
        .copy_from_slice(&source_character_header_data);

    // Actiate target slot
    new_save[save_model::CHAR_ACTIVE_STATUS_START_INDEX + target_character_slot] = 1;

    let mut md5 = md5::Context::new();

    md5.consume(source_character_save_data);
    let slot_checksum_digest = md5.compute();
    let slot_checksum = slot_checksum_digest.as_slice();
    new_save[save_model::get_slot_start_position(target_character_slot) - 0x10..][..0x10]
        .copy_from_slice(slot_checksum);

    // reset hasher
    let mut md5 = md5::Context::new();
    md5.consume(
        &new_save[save_model::SAVE_HEADERS_SECTION_START_INDEX..]
            [..save_model::SAVE_HEADERS_SECTION_LENGTH],
    );
    let header_checksum_digest = md5.compute();
    let header_checksum = header_checksum_digest.as_slice();
    new_save[save_model::SAVE_HEADERS_SECTION_START_INDEX - 0x10..][..0x10]
        .copy_from_slice(header_checksum);

    Ok(new_save)
}

File: src/file_io.go
package main

import (
	"bytes"
	"crypto/md5"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

const MaxCharacters = 10

func ListActiveCharacters(data []byte) []Character {
	out := make([]Character, MaxCharacters)
	n := 0
	for i := range MaxCharacters {
		if c := NewActiveCharacter(data, i); c != nil {
			out[n] = *c
			n++
		}
	}
	return out[:n]
}

func ListCharacters(data []byte) []*Character {
	out := make([]*Character, MaxCharacters)
	for i := range MaxCharacters {
		out[i] = NewActiveCharacter(data, i)
	}
	return out
}

func ListAllCharacters(data []byte) []Character {
	out := make([]Character, MaxCharacters)
	for i := range MaxCharacters {
		out[i] = NewCharacter(data, i)
	}
	return out
}

func unixTimestamp() uint64 {
	return uint64(time.Now().Unix())
}

func WriteBackup(data []byte, destination *string) (string, error) {
	var base string
	if destination != nil {
		base = *destination
	} else {
		dir, err := os.UserHomeDir()
		if err != nil {
			return "", errors.New("unable to determine home directory")
		}
		base = filepath.Join(dir, "Documents", "er_mule_copier_backups")
		if err := os.MkdirAll(base, 0755); err != nil {
			return "", err
		}
	}

	filename := fmt.Sprintf("ER0000 backup from %d.sl2", unixTimestamp())
	fullpath := filepath.Join(base, filename)

	if err := os.WriteFile(fullpath, data, 0644); err != nil {
		return "", err
	}
	return fullpath, nil
}

func WriteFile(data []byte, filename string) error {
	backup := filename + ".bak"
	_ = os.Remove(backup) // best-effort, matches Rust behavior
	return os.WriteFile(filename, data, 0644)
}

func subslicePositions(needle, haystack []byte) []int {
	if len(needle) == 0 || len(haystack) < len(needle) {
		return nil
	}

	// worst case: every byte matches
	out := make([]int, 0, 16)
	for i := 0; i <= len(haystack)-len(needle); i++ {
		if bytes.Equal(haystack[i:i+len(needle)], needle) {
			out = append(out, i)
		}
	}
	return out
}

func GenerateNewData(
	sourceData []byte,
	sourceSlot int,
	targetData []byte,
	targetSlot int,
) ([]byte, error) {

	newsave := make([]byte, len(targetData))
	copy(newsave, targetData)

	sourceID, err := ParseSteamID(sourceData)
	if err != nil {
		return nil, err
	}
	targetID, err := ParseSteamID(targetData)
	if err != nil {
		return nil, err
	}

	sourceSlotData := ParseSaveData(sourceData, sourceSlot)
	sourceHeaderData := ParseHeaderData(sourceData, sourceSlot)

	for _, pos := range subslicePositions(sourceID[:], sourceSlotData) {
		copy(
			sourceSlotData[pos:pos+STEAM_ID_LENGTH],
			targetID[:],
		)
	}

	copy(
		newsave[GetSlotStartPosition(targetSlot):GetSlotStartPosition(targetSlot)+SLOT_LENGTH],
		sourceSlotData,
	)

	copy(
		newsave[GetHeaderStartPosition(targetSlot):GetHeaderStartPosition(targetSlot)+SAVE_HEADER_LENGTH],
		sourceHeaderData,
	)

	newsave[CHAR_ACTIVE_STATUS_START_INDEX+targetSlot] = 1

	slotHash := md5.Sum(sourceSlotData)
	copy(
		newsave[GetSlotStartPosition(targetSlot)-0x10:GetSlotStartPosition(targetSlot)],
		slotHash[:],
	)

	headerHash := md5.Sum(
		newsave[SAVE_HEADERS_SECTION_START_INDEX : SAVE_HEADERS_SECTION_START_INDEX+SAVE_HEADERS_SECTION_LENGTH],
	)

	copy(
		newsave[SAVE_HEADERS_SECTION_START_INDEX-0x10:SAVE_HEADERS_SECTION_START_INDEX],
		headerHash[:],
	)

	return newsave, nil
}
